src/chain/chain-supervision.cc

```diff
@@ -207,6 +209,47 @@ bool PhoneLatticeToProtoSupervision(const SupervisionOptions &opts,
   return true;
 }
 
+bool NormalizePhoneLattice(CompactLattice *clat) {
+  if (clat->Properties(fst::kTopSorted, false) == 0) {
+    if (fst::TopSort(clat) == false) {
+      KALDI_WARN << "Cycles detected in lattice: cannot normalize.";
+      return false;
+    }
+  }
+
+  std::vector<double> beta;
+  if (!ComputeCompactLatticeBetas(*clat, &beta)) {
+    KALDI_WARN << "Failed to compute backward probabilities on lattice.";
+    return false;
+  }
+
+  CompactLattice::Arc::StateId start = clat->Start();  // Should be 0
+  BaseFloat total_backward_cost = beta[start];
+
+  for (fst::StateIterator<CompactLattice> sit(*clat); !sit.Done(); sit.Next()) {
+    CompactLatticeWeight f = clat->Final(sit.Value());
+    LatticeWeight w = f.Weight();
+    w.SetValue1(w.Value1() + total_backward_cost);
```

Comment:
So do you mean I should convert it to FST, do the normalization, and the convert it back to lattice (all before calling PhoneLatticeToProtoSupervisionInternal) or alternatively do it on the fst of the resulting ProtoSupervision (i.e. after calling PhoneLatticeToProtoSupervisionInternal)?